ui: restore --delete indicates number of deleted files

konidev20 opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------
* adds new field `FilesDeleted` to the State struct, JSON and text progress updaters
* renamed `ReportDeletedFile` to `ReportDeletion`
* increment FilesDeleted count when ReportedDeletion is called

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Closes #5092 

Checklist
---------
- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

konidev20 committed:
ui: restore --delete indicates number of deleted files

* adds new field `FilesDeleted` to the State struct, JSON and text progress updaters
* increment FilesDeleted count when ReportedDeletedFile

MichaelEischer commented on internal/ui/restore/progress.go:
```diff
@@ -129,6 +130,8 @@ func (p *Progress) ReportDeletedFile(name string) {
 		return
 	}
 
+	p.s.FilesDeleted++
```
Files are deleted using `fs.RemoveAll` which means that a single name can actually refer to a full folder of files. It feels wrong to count it as a single item in that case.

konidev20 replied:
@MichaelEischer, I see. I checked the logic, I have modified this place to do a file traversal to count the files, run the `fs.RemoveAll`, and then finally update the count of removed files. Is there something I might still be missing?

konidev20 committed:
ui: collect the files to be deleted, delete, then update the count post deletion

MichaelEischer commented on internal/restorer/restorer.go:
```diff
@@ -511,12 +511,32 @@ func (res *Restorer) removeUnexpectedFiles(ctx context.Context, target, location
+				if err != nil {
+					return err
+				}
+				if !info.IsDir() {
```
`ReportDeletedFile` previously also reported deletions of directories. I think we should keep that behavior. That is, just collect all directory entries that would be deleted. And ideally rename `ReportDeletedFile` to `ReportDeletion`.

konidev20 replied:
Done. 
1. I have removed the directory check. 
2. I have renamed the function to `ReportDeletion`

MichaelEischer reviewed (commented):
I have only one small nit remaining for the code. Besides that, please add a short changelog entry.

MichaelEischer commented on internal/ui/restore/json.go:
```diff
@@ -108,6 +110,7 @@ type statusUpdate struct {
 	TotalFiles     uint64  `json:"total_files,omitempty"`
 	FilesRestored  uint64  `json:"files_restored,omitempty"`
 	FilesSkipped   uint64  `json:"files_skipped,omitempty"`
+	FilesDeleted   uint64  `json:"files_deleted,omitempty"`
```
The new field must be documented at `doc/075_scripting.rst`.

konidev20 replied:
I have updated the documentation with the `files_deleted` json field.

konidev20 committed:
docs: update scripting output fields for restore command

konidev20 committed:
This is a combination of 3 commits.

ui: report deleted directories and refactor function name to ReportDeletion

changelog: issue-5092

lint: fix unused variable

MichaelEischer reviewed (approved):
LGTM. Thanks!
